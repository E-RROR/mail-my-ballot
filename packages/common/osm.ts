import { Address } from "."
import { isState, State } from "./states"

/*
  In theory, this module could exist on the client, but we need it in common for
  testing on the server.  See `contact/index.test.ts`.
*/

const fetch = (typeof window !== 'undefined' && window.fetch) ? window.fetch : require('node-fetch')

const getJson = async <T>(url: string): Promise<T> => {
  const result = await fetch(url)
  return await result.json()
}

interface BasicAddress {
  houseNumber: string
  road: string
  unit: string
  city: string
  country: string
  state: string
  postcode: string
  county: string
}

type GenericRecord = Record<string, unknown>

const osmQuery = async <T>(
  query: string,
  parseSummary: (summary: GenericRecord) => T | null,
  parseDetail?: (detail: GenericRecord) => T | null,
): Promise<T | null> => {
  const summary = (await getJson<Array<GenericRecord>>(`https://nominatim.openstreetmap.org/search/${query}?format=json&countrycodes=us`))[0]
  if (!summary) return null

  const summaryResult = parseSummary(summary)
  if (summaryResult) return summaryResult
  
  if (!parseDetail) return null
  const osmId = (summary.osm_type as string)[0].toUpperCase() + summary.osm_id
  const detail = (await getJson<Array<GenericRecord>>(`https://nominatim.openstreetmap.org/lookup?osm_ids=${osmId}&format=json`))[0]
  if (!detail) return null
  const detailResult = parseDetail(detail)
  return detailResult
}

class PartsParser {
  parts: string[]
  displayName: string

  constructor(displayName: string) {
    this.displayName = displayName
    this.parts = displayName.split(',').map(x => x.trim()).reverse()
  }

  parsePart(re?: RegExp | ((part: string) => boolean)): string | undefined {
    const match = (
      (typeof re === 'undefined') ? (_: string) => true : 
      (typeof re === 'function') ? re :
      (part: string) => !!part.match(re)
    )
    if (this.parts[0] && match(this.parts[0])) {
      return (this.parts.shift() as string)
    }
    return undefined
  }

  parse() {
    const country = this.parsePart(/^United States of America$/)
    const postcode = this.parsePart(/^\d{5}/)
    const state = this.parsePart(isState) as State | null
    const county = this.parsePart(/County$/)
    const city = this.parsePart()

    return {
      country,
      postcode,
      state,
      county,
      city,
    }
  }
}

export const fetchState = async (zip: string): Promise<State | null> => {
  const query = `${zip} United States`
  return osmQuery(query, (summary) => {
    const parser = new PartsParser(summary.display_name as string)
    return parser.parse().state || null
  })
}

const parseDisplayName = (displayName: string, unit: string): BasicAddress | null => {
  /*
  Parse a display_name field from osm sumary result, e.g.
  2125, Butterfield Road, Regents Club Of Troy, Big Beaver, Troy, Oakland County, Michigan, 48084, United States of America
  */

  // 
  const parts = displayName.split(',')
  if(parts.length < 7) return null
  return {
    houseNumber: parts[0].trim(),
    road: parts[1].trim(),
    unit,
    city: parts[parts.length - 5].trim(),
    county: parts[parts.length - 4].trim(),
    state: parts[parts.length - 3].trim(),
    postcode: parts[parts.length - 2].trim(),
    country: parts[parts.length - 1].trim(),
  }
}

export const geocode = async (queryAddr: string, unit = ''): Promise<Address | null> => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const summary = (await getJson<Array<any>>(`https://nominatim.openstreetmap.org/search/${queryAddr}?format=json&countrycodes=us`))[0]
  if (!summary) {
    return null
  }
  const osmId = summary.osm_type[0].toUpperCase() + summary.osm_id
  const summaryAddress = parseDisplayName(summary.display_name, unit)
  if (summaryAddress) {
    return {
      queryAddr,
      fullAddr: summary.display_name,
      ...summaryAddress
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const detail = (await getJson<Array<any>>(`https://nominatim.openstreetmap.org/lookup?osm_ids=${osmId}&format=json`))[0]
  if (!detail) {
    return null
  }
  const { address } = detail
  const {
    house_number: rawHouseNumber,
    road,
    city,
    state,
    postcode,
    country,
    county
  } = address

  // OSM is often missing the house number on response.  So we'll infer it
  const fullAddr = summary.display_name
  const houseNumber = rawHouseNumber ? rawHouseNumber : fullAddr.split(',')[0].trim()

  return {
    queryAddr,
    fullAddr,
    houseNumber,
    road,
    unit,
    city,
    state,
    postcode,
    country,
    county,
  }
}
