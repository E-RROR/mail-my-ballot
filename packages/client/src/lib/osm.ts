import { Address } from "../common"

const getJson = async <T>(url: string): Promise<T> => {
  const result = await fetch(url)
  return await result.json()
}

interface BasicAddress {
  houseNumber: string
  road: string
  unit: string
  city: string
  country: string
  state: string
  postcode: string
  county: string
}

const parseDisplayName = (displayName: string, unit: string): BasicAddress | null => {
  /*
  Parse a display_name field from osm sumary result, e.g.
  2125, Butterfield Road, Regents Club Of Troy, Big Beaver, Troy, Oakland County, Michigan, 48084, United States of America
  */

  // 
  const parts = displayName.split(',')
  if(parts.length < 7) return null
  return {
    houseNumber: parts[0].trim(),
    road: parts[1].trim(),
    unit,
    city: parts[parts.length - 5].trim(),
    county: parts[parts.length - 4].trim(),
    state: parts[parts.length - 3].trim(),
    postcode: parts[parts.length - 2].trim(),
    country: parts[parts.length - 1].trim(),
  }
}

export const geocode = async (queryAddr: string, unit: string): Promise<Address | null> => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const summary = (await getJson<Array<any>>(`https://nominatim.openstreetmap.org/search/${queryAddr}?format=json&countrycodes=us`))[0]
  if (!summary) {
    return null
  }
  const osmId = summary.osm_type[0].toUpperCase() + summary.osm_id
  const summaryAddress = parseDisplayName(summary.display_name, unit)
  if (summaryAddress) {
    return {
      queryAddr,
      fullAddr: summary.display_name,
      ...summaryAddress
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const detail = (await getJson<Array<any>>(`https://nominatim.openstreetmap.org/lookup?osm_ids=${osmId}&format=json`))[0]
  if (!detail) {
    return null
  }
  const { address } = detail
  const {
    house_number: rawHouseNumber,
    road,
    city,
    state,
    postcode,
    country,
    county
  } = address

  // OSM is often missing the house number on response.  So we'll infer it
  const fullAddr = summary.display_name
  const houseNumber = rawHouseNumber ? rawHouseNumber : fullAddr.split(',')[0].trim()

  return {
    queryAddr,
    fullAddr,
    houseNumber,
    road,
    unit,
    city,
    state,
    postcode,
    country,
    county,
  }
}
