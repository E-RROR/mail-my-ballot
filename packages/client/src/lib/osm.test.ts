import { osmGeocode } from './osm'

test('OSM is returning stable results', async () => {
  const testCases: [string, string][] = [
    ['100 S Biscayne Blvd, Miami, FL 33131', 'Miami-Dade County, Florida'],
    ['5443 Main St, Port Richey, FL 34652', 'Pasco County, Florida'],
    ['2000 W Commercial Blvd, Fort Lauderdale, FL 33309', 'Broward County, Florida'],
  ]

  try {
    await Promise.all(
      testCases.map(async (pair) => {
        const [addr, county] = pair
        const result = await osmGeocode(addr, '1A')
        expect(result).toMatchSnapshot(county);
      })
    )
  } catch(e) {
    fail();
  }
})
