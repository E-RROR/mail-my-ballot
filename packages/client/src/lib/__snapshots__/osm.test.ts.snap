// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OSM is returning stable results:  Berrien County, Michigan 1`] = `
Object {
  "city": "Saint Joseph",
  "country": "United States of America",
  "county": "Berrien County",
  "fullAddr": "700, Broad Street, Saint Joseph, Berrien County, Michigan, 49085, United States of America",
  "houseNumber": "700",
  "postcode": "49085",
  "queryAddr": "700 Broad Street St Joseph, MI 49085",
  "road": "Broad Street",
  "state": "Michigan",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Broward County, Florida 1`] = `
Object {
  "city": "Fort Lauderdale",
  "country": "United States of America",
  "county": "Broward County",
  "fullAddr": "2000, West Commercial Boulevard, Fort Lauderdale, Broward County, Florida, 33309, United States of America",
  "houseNumber": "2000",
  "postcode": "33309",
  "queryAddr": "2000 W Commercial Blvd, Fort Lauderdale, FL 33309",
  "road": "West Commercial Boulevard",
  "state": "Florida",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Canton, Wayne County, Michigan 1`] = `
Object {
  "city": "Canton",
  "country": "United States of America",
  "county": "Wayne County",
  "fullAddr": "45525, Hanford Road, Cherry Hill, Canton, Wayne County, Michigan, 48187, United States of America",
  "houseNumber": "45525",
  "postcode": "48187",
  "queryAddr": "45525 Hanford Rd, Canton, MI 48187",
  "road": "Hanford Road",
  "state": "Michigan",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Dearborn Heights, Wayne County, Michigan 1`] = `
Object {
  "city": "Dearborn Heights",
  "country": "United States of America",
  "county": "Wayne County",
  "fullAddr": "24624, West Warren Street, Dearborn Heights, Wayne County, Michigan, 48127, United States of America",
  "houseNumber": "24624",
  "postcode": "48127",
  "queryAddr": "24624 W Warren St, Dearborn Heights, MI",
  "road": "West Warren Street",
  "state": "Michigan",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Dearborn, Wayne County, Michigan 1`] = `
Object {
  "city": "Dearborn",
  "country": "United States of America",
  "county": "Wayne County",
  "fullAddr": "22100, Michigan Avenue, Dearborn, Wayne County, Michigan, 48124, United States of America",
  "houseNumber": "22100",
  "postcode": "48124",
  "queryAddr": "22100 Michigan Ave. Dearborn, MI 48124",
  "road": "Michigan Avenue",
  "state": "Michigan",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Miami-Dade County, Florida 1`] = `
Object {
  "city": "Miami",
  "country": "United States of America",
  "county": "Miami-Dade County",
  "fullAddr": "100, Biscayne Boulevard, The Roads, Miami, Miami-Dade County, Florida, 33131, United States of America",
  "houseNumber": "100",
  "postcode": "33131",
  "queryAddr": "100 S Biscayne Blvd, Miami, FL 33131",
  "road": "Biscayne Boulevard",
  "state": "Florida",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Muskegon County, Michigan 1`] = `
Object {
  "city": "Muskegon Heights",
  "country": "United States of America",
  "county": "Muskegon County",
  "fullAddr": "2724, Peck Street, Muskegon Heights, Muskegon County, Michigan, 49444, United States of America",
  "houseNumber": "2724",
  "postcode": "49444",
  "queryAddr": "2724 Peck St, Muskegon Heights, MI 49444",
  "road": "Peck Street",
  "state": "Michigan",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Pasco County, Florida 1`] = `
Object {
  "city": "New Port Richey",
  "country": "United States of America",
  "county": "Pasco County",
  "fullAddr": "5443, Main Street, New Port Richey, Pasco County, Florida, 34652, United States of America",
  "houseNumber": "5443",
  "postcode": "34652",
  "queryAddr": "5443 Main St, Port Richey, FL 34652",
  "road": "Main Street",
  "state": "Florida",
  "unit": "1A",
}
`;
